pyron-plugin/src/main/scala/com/cloudentity/pyron/plugin/util/value/ValueResolver.scala

```diff
@@ -1,54 +1,104 @@
+      case HeaderRef(header, FirstHeaderRefType) => req.request.headers.get(header) // take first header value
+      case AuthnRef(path) => extractAuthnCtxAttribute(req.authnCtx, path).flatMap(_.asString)
     }
 
-  def resolveString(req: RequestCtx, valueOrRef: ValueOrRef): Option[String] =
-    valueOrRef match {
-      case BodyRef(_) => resolveString(req, req.request.bodyOpt.flatMap(buf => Try(buf.toJsonObject).toOption), valueOrRef)
-      case x          => resolveString(req, None, x)
-    }
+  def resolveString(req: RequestCtx, valueOrRef: ValueOrRef): Option[String] = valueOrRef match {
+    case BodyRef(_) => resolveString(req, req.request.bodyOpt.flatMap(buf => Try(buf.toJsonObject).toOption), valueOrRef)
+    case x          => resolveString(req, None, x)
+  }
 
   def resolveListOfStrings(req: RequestCtx, bodyOpt: Option[JsonObject], valueOrRef: ValueOrRef): Option[List[String]] =
-    valueOrRef match {
+    (valueOrRef: @unchecked) match { // no default case, fail fast
       case Value(value)                          =>
         if (value.asObject.nonEmpty) circeJsonDynamicString(req, bodyOpt, value)
         else circeJsonToJsonValue(value).asListOfStrings
-      case BodyRef(path)                         => bodyOpt.flatMap(extractBodyAttribute(_, path)).flatMap(_.asListOfStrings)
-      case PathParamRef(param)                   => req.request.uri.pathParams.value.get(param).map(List(_))
-      case AuthnRef(path)                        => extractAuthnCtxAttribute(req.authnCtx, path).flatMap(_.asListOfStrings)
+      case HostRef                               => resolveHost(req).map(List(_))
+      case HostNameRef                           => resolveHostName(req).map(List(_))
+      case HostPortRef                           => resolveHostPort(req).map(List(_))
+      case SchemeRef                             => resolveScheme(req).map(List(_))
+      case LocalHostRef                          => resolveLocalHost(req).map(List(_))
+      case RemoteHostRef                         => resolveRemoteHost(req).map(List(_))
+      case CookieRef(cookie)                     => resolveCookie(req, cookie).map(List(_))
+      case BodyRef(path)                         => resolveBody(bodyOpt, path).flatMap(_.asListOfStrings)
+      case PathParamRef(param)                   => resolvePathParam(req, param).map(List(_))
+      case QueryParamRef(param)                  => req.request.uri.query.getValues(param)
       case HeaderRef(header, FirstHeaderRefType) => req.request.headers.get(header).map(List(_))
       case HeaderRef(header, AllHeaderRefType)   => req.request.headers.getValues(header)
+      case AuthnRef(path)                        => extractAuthnCtxAttribute(req.authnCtx, path).flatMap(_.asListOfStrings)
     }
 
   def resolveJson(req: RequestCtx, bodyOpt: Option[JsonObject], valueOrRef: ValueOrRef): Option[JsonValue] =
-    valueOrRef match {
+    (valueOrRef: @unchecked) match { // no default case, fail fast
```

Comment:
Same as above. Let's call compiler for the rescue ;)

Reply:
same as above also :)

Reply:
My bad, please see my request in TransformRequestPlugin.

Reply:
done

Reply:
done